src/data_processor.py

```diff
@@ -1,6 +1,160 @@
+            if not isinstance(dictionary[user][metric], str):
+                # if list, add the length of the list to new dictionary
+                if isinstance(dictionary[user][metric], list):
+                    amount = len(dictionary[user][metric])
+                    # print(amount)
+                    if metric not in category_scores:
+                        category_scores[metric] = [amount]
+                    else:
+                        category_scores[metric].append(amount)
+                # if not list, add values to new dictionary
+                else:
+                    if metric not in category_scores:
+                        category_scores[metric] = [dictionary[user][metric]]
+                    else:
+                        category_scores[metric].append(dictionary[user][metric])
+
+    # return the new dictionary
+    return category_scores
+
+
+def calculate_iqr_score(data_list, below_weight, above_weight, within_weight):
+    """Calculate a team score with interquartile range."""
+    # pylint: disable=too-many-locals
+    below_amount = 0
+    above_amount = 0
+    within_amount = 0
+    dataset = data_list
+    # calculate the size of the dataset
+    size = len(dataset)
+    # sort the dataset in ascending order
+    sorted(dataset)
+    # determine the first and third quartiles
+    q1, q3 = np.percentile(dataset, [25, 75])
+    # calculate the interquartile range (q3-q1)
+    iqr = q3 - q1
+    # calculate the lower bound of the IQR
+    lower_bound = q1 - (1.5 * iqr)
+    # calculate the upper bound of the IQR
+    upper_bound = q3 + (1.5 * iqr)
+    # for any value above, below or within the IQR add to the total
+    for d in dataset:
+        if d < lower_bound:
+            below_amount = below_amount + 1
+        if d > upper_bound:
+            above_amount = above_amount + 1
+        if lower_bound <= d <= upper_bound:
+            within_amount = within_amount + 1
+    # create factions of above, below, and within IQR amounts (divide by size) and round
+    # pylint: disable=round-builtin
+    below_fraction = round(below_amount / size, 2)
+    above_fraction = round(above_amount / size, 2)
+    within_fraction = round(within_amount / size, 2)
+    # calculate all areas with their weight measurement
+    weighted_below = round(below_weight * below_fraction, 2)
+    weighted_above = round(above_weight * above_fraction, 2)
+    weighted_within = round(within_weight * within_fraction, 2)
+    # add weighted scores together to calcuate overall team score as percentage
+    category_score = (weighted_below + weighted_above + weighted_within) * 100
+
+    return category_score
+
+
+def calculate_team_score(dictionary, below_weight, above_weight, within_weight):
+    """Calculate the average team score."""
+    # create a new dictionary and assign it to the dictionary iteration
+    metrics_dictionary = iterate_nested_dictionary(dictionary)
```

Comment:
Can you please explain a bit more about what this does in your comments.

Reply:
I added comments in both the `iterate_nested_dictionary` and `calculate_team_score` functions to clarify the purpose. Essentially the nested dictionary is iterated through, and all values for a specific metric ("COMMITS, "ADDED" etc) are added to a list, which is the value of a dictionary with its metric as the key. This allows much easier iqr score calculation for each category.